[Build] Build host libraries into toolchain's lib/swift/host as shared libraries

DougGregor opened the pull request:
Start building shared libraries into the toolchain within `lib/swift/host`:
* Move `swift_CompilerPluginSupport` into `lib/swift/host`
* Start building swift-syntax as resilient shared libraries into `lib/swift/host`
* Have the compiler, SourceKit, and other tools link against these shared libraries

This means we won't get redundant copies of the swift-syntax libraries across the compiler and SourceKit (and other executables), and that compiler plugins can build against these shared versions safely.

DougGregor committed:
[CMake] Link to swift-syntax libriares from lib/swift/host.

DougGregor committed:
[Build] Start installing the "early" Swift Syntax into the toolchain

DougGregor committed:
Start building libswift_CompilerPluginSupport into lib/swift/host.

DougGregor committed:
[Macros] Fix path to host libraries in macro tests

DougGregor committed:
Ensure that the host library path is provided to tests.

Eliminate the "copy the plugin library" step for these tests, which was
always a hack and doesn't scale when we add more shared libraries.

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please smoke test

DougGregor committed:
Build swift-syntax as shared libraries instead of static libraries.

This allows the various binaries (swift-frontend, SourceKit, etc.) to
share the same code, as well as allowing plugins to link against these
shared libraries.

DougGregor committed:
[lit] Work around issue where can't set the host library environment.

DougGregor committed:
Copy all of the swift-syntax Swift modules.

Make the build directory for lib/swift/host look like the eventually
toolchain directory by copying in all of the Swift module files. This
makes it easier to correctly build anything requiring these libraries
against either the build or the install directories.

DougGregor committed:
Fix Python fun with strings vs. lists of strings.

DougGregor committed:
Fix up a few more tests that need the host library path

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please smoke test

compnerd commented on lib/CompilerPluginSupport/CMakeLists.txt:
```diff
@@ -18,6 +18,12 @@ if(SWIFT_SWIFT_PARSER)
+    PROPERTIES
+      ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib/swift/host"
+      LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib/swift/host"
+  )
```
Please add `RUNTIME_OUTPUT_DIRECTORY` as well.  I don't think that we can use `lib/swift/host` on Windows.  The general convention there is that the `RUNTIME_OUTPUT_DIRECTORY` is `bin` and the `LIBRARY_OUTPUT_DIRECTORY` and `ARCHIVE_OUTPUT_DIRECTORY` is `lib`.  That puts the import library (the moral equivalent of TBDs) and static libraries into `lib` and DLLs into `bin` where it will be found when doing the library lookup as there is no concept of RPATH.

DougGregor replied:
(I see how it's used elsewhere, never mind. I'll keep poking at this)

compnerd commented on lib/CompilerPluginSupport/CMakeLists.txt:
```diff
@@ -82,17 +88,17 @@ if(SWIFT_SWIFT_PARSER)
-      DESTINATION "lib${LLVM_LIBDIR_SUFFIX}"
+      DESTINATION "lib${LLVM_LIBDIR_SUFFIX}/swift/host"
       COMPONENT compiler)
 
```
Similarly, we need `RUNTIME DESTINATION` here.

DougGregor replied:
Alright, we'll need to do this over in swift-syntax as well. I'd actually rather start there, because `swift_CompilerPluginSupport` might not stick around.

compnerd reviewed (commented):
This is going to need a complementary change in swift-installer-scripts to update the packaging for ARM64 and AMD64.  Please make sure to run a windows packaging test (build toolchain windows platform).  We really need to get a smoke test setup for that as it will not identify if there are missing libraries :-(.

DougGregor committed:
Set host libraries rpath correctly for bootstrapped builds

DougGregor committed:
[lit] Use a more appropriate default for host_library_env

DougGregor commented:
> This is going to need a complementary change in swift-installer-scripts to update the packaging for ARM64 and AMD64. Please make sure to run a windows packaging test (build toolchain windows platform). We really need to get a smoke test setup for that as it will not identify if there are missing libraries :-(.

I'll take a look!

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please smoke test

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please build toolchain windows platform

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please build toolchain

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please clean smoke test macOS

DougGregor committed:
[Build] Make sure to install SwiftSyntax into the toolchain

... otherwise we build a broken compiler in the toolchain, because the
shared libraries for SwiftSyntax aren't present.

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please clean smoke test

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please build toolchain

DougGregor committed:
Only install SwiftSyntax when we can build it

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please clean smoke test

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please build toolchain

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please clean test macOS

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please clean smoke test macOS

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please smoke test macOS

DougGregor committed:
Make sure that the _InternalSwiftScan library can find host dylibs

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please smoke test

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please build toolchain

DougGregor committed:
[CMake] Add rpath for the installed location of lib_InternalSwiftScan.

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please smoke test

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please smoke test

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please build toolchain

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please clean smoke test macOS

DougGregor commented:
https://github.com/apple/swift-syntax/pull/1097

@swift-ci please smoke test macOS
